compiler.go

```diff
@@ -385,12 +403,30 @@ func (t *task) asFile(ctx context.Context, name string, r SearchResult) (linker.
 	}
 
 	var deps []linker.File
-	if len(parseRes.FileDescriptorProto().Dependency) > 0 {
-		t.r.setBlockedOn(parseRes.FileDescriptorProto().Dependency)
+	fileDescriptorProto := parseRes.FileDescriptorProto()
+	var wantsDescriptorProto bool
+	imports := fileDescriptorProto.Dependency
+
+	if t.e.hasOverrideDescriptorProto() {
+		// we only consider implicitly including descriptor.proto if it's overidden
+		includesDescriptorProto := name == descriptorProtoPath
+		for _, dep := range fileDescriptorProto.Dependency {
+			if !includesDescriptorProto && dep == descriptorProtoPath {
+				includesDescriptorProto = true
+			}
+		}
+		if !includesDescriptorProto {
+			wantsDescriptorProto = true
+			imports = append(imports, descriptorProtoPath)
```

Comment:
Don't we want to ensure we create a new slice here so that if imports has capacity, we don't change the underlying FileDescriptorProto.Dependency?

Reply:
Yes, that's a good point. Otherwise, anything sharing the backing array with this slice will get erroneously mutated.

Reply:
done